## Se intentó diseñar un entorno profesional de trabajo en VSC, con ayuda de la AI, para que quede como modelo de futuros proyectos (más tarde todos los detalles quedarán reflejados en el README.MD).

In [2]:
import os
import pandas as pd
from dotenv import load_dotenv

# Cargar las rutas definidas en el archivo .env
load_dotenv()

clientes_path = os.getenv("CLIENTES_PATH")
marketing_path = os.getenv("MARKETING_PATH")
ventas_path = os.getenv("VENTAS_PATH")

print("Rutas configuradas correctamente:")
print("Clientes :", clientes_path)
print("Marketing:", marketing_path)
print("Ventas   :", ventas_path)

Rutas configuradas correctamente:
Clientes : data/clientes.csv
Marketing: data/marketing.csv
Ventas   : data/ventas.csv


In [3]:
import os
from pathlib import Path
import pandas as pd
from dotenv import load_dotenv

# Carga las variables definidas en el archivo .env ubicado en la raíz
load_dotenv()

# Obtenemos las rutas relativas de los datasets
clientes_path = os.getenv("CLIENTES_PATH", "data/clientes.csv")
marketing_path = os.getenv("MARKETING_PATH", "data/marketing.csv")
ventas_path = os.getenv("VENTAS_PATH", "data/ventas.csv")

print("Rutas configuradas:")
print("Clientes :", clientes_path)
print("Marketing:", marketing_path)
print("Ventas   :", ventas_path)

Rutas configuradas:
Clientes : data/clientes.csv
Marketing: data/marketing.csv
Ventas   : data/ventas.csv


## Cargar Archivos

In [27]:
import pandas as pd

df_clientes = pd.read_csv("data/clientes.csv")
df_marketing = pd.read_csv("data/marketing.csv")
df_ventas = pd.read_csv("data/ventas.csv")

print("Archivos cargados")

Archivos cargados


# Fase Exploración inicial de datos ( Análisis Exploratorio de Datos preliminar / Diagnóstico)

## 5 filas de clientes

In [ ]:

df_clientes.head()

,id_cliente,nombre,edad,ciudad,ingresos
0,1,Aloysia Screase,44,Mar del Plata,42294.68
1,2,Kristina Scaplehorn,25,Posadas,24735.04
2,3,Filip Castagne,50,Resistencia,35744.85
3,4,Liuka Luard,39,Bahía Blanca,27647.96
4,5,Dore Cockshtt,28,Rosario,28245.65


## Ver dimensiones (filas, columnas) y tipos de datos de clientes

In [ ]:

print("Dimensiones de df_clientes:", df_clientes.shape)
df_clientes.info()

Dimensiones de df_clientes: (567, 5)
<class 'pandas.DataFrame'>
RangeIndex: 567 entries, 0 to 566
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id_cliente  567 non-null    int64  
 1   nombre      567 non-null    str    
 2   edad        567 non-null    int64  
 3   ciudad      567 non-null    str    
 4   ingresos    567 non-null    float64
dtypes: float64(1), int64(2), str(2)
memory usage: 22.3 KB


## Inspeccionar las primeras filas y estructura de marketing

In [ ]:

display(df_marketing.head())
print("Dimensiones de df_marketing:", df_marketing.shape)
df_marketing.info()

,id_campanha,producto,canal,costo,fecha_inicio,fecha_fin
0,74,Adorno de pared,TV,4.81,20/03/2024,03/05/2024
1,12,Tablet,RRSS,3.40,26/03/2024,13/05/2024
2,32,Lámpara de mesa,Email,5.54,28/03/2024,20/04/2024
3,21,Smartphone,RRSS,6.37,29/03/2024,16/05/2024
4,58,Alfombra,Email,4.25,31/03/2024,05/05/2024


Dimensiones de df_marketing: (90, 6)
<class 'pandas.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id_campanha   90 non-null     int64  
 1   producto      90 non-null     str    
 2   canal         90 non-null     str    
 3   costo         90 non-null     float64
 4   fecha_inicio  90 non-null     str    
 5   fecha_fin     90 non-null     str    
dtypes: float64(1), int64(1), str(4)
memory usage: 4.3 KB


## Inspeccionar las primeras filas y estructura de ventas

In [ ]:

display(df_ventas.head())
print("Dimensiones de df_ventas:", df_ventas.shape)
df_ventas.info()

,id_venta,producto,precio,cantidad,fecha_venta,categoria
0,792,Cuadro decorativo,$69.94,5.0,02/01/2024,Decoración
1,811,Lámpara de mesa,$105.10,5.0,02/01/2024,Decoración
2,1156,Secadora,$97.96,3.0,02/01/2024,Electrodomésticos
3,1372,Heladera,$114.35,8.0,02/01/2024,Electrodomésticos
4,1546,Secadora,$106.21,4.0,02/01/2024,Electrodomésticos


Dimensiones de df_ventas: (3035, 6)
<class 'pandas.DataFrame'>
RangeIndex: 3035 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3035 non-null   int64  
 1   producto     3035 non-null   str    
 2   precio       3033 non-null   str    
 3   cantidad     3033 non-null   float64
 4   fecha_venta  3035 non-null   str    
 5   categoria    3035 non-null   str    
dtypes: float64(1), int64(1), str(4)
memory usage: 142.4 KB


## Contar valores nulos por columna en cada dataset

In [ ]:

print("--- Nulos en Clientes ---")
print(df_clientes.isna().sum())

print("\n--- Nulos en Marketing ---")
print(df_marketing.isna().sum())

print("\n--- Nulos en Ventas ---")
print(df_ventas.isna().sum())

--- Nulos en Clientes ---
id_cliente    0
nombre        0
edad          0
ciudad        0
ingresos      0
dtype: int64

--- Nulos en Marketing ---
id_campanha     0
producto        0
canal           0
costo           0
fecha_inicio    0
fecha_fin       0
dtype: int64

--- Nulos en Ventas ---
id_venta       0
producto       0
precio         2
cantidad       2
fecha_venta    0
categoria      0
dtype: int64


## Contar filas completas duplicadas en cada dataset

In [ ]:

print("Filas duplicadas en Clientes :", df_clientes.duplicated().sum())
print("Filas duplicadas en Marketing:", df_marketing.duplicated().sum())
print("Filas duplicadas en Ventas   :", df_ventas.duplicated().sum())

Filas duplicadas en Clientes : 0
Filas duplicadas en Marketing: 0
Filas duplicadas en Ventas   : 35


## Ver algunos ejemplos de las filas duplicadas en ventas

In [ ]:

filas_duplicadas = df_ventas[df_ventas.duplicated(keep=False)]
filas_duplicadas.sort_values(by=list(df_ventas.columns)).head(6)

,id_venta,producto,precio,cantidad,fecha_venta,categoria
820,56,Cortinas,$66.24,5.0,05/04/2024,Decoración
826,56,Cortinas,$66.24,5.0,05/04/2024,Decoración
832,145,Alfombra,$41.72,8.0,06/04/2024,Decoración
844,145,Alfombra,$41.72,8.0,06/04/2024,Decoración
878,204,Jarrón decorativo,$94.75,1.0,08/04/2024,Decoración
884,204,Jarrón decorativo,$94.75,1.0,08/04/2024,Decoración


## Ayuda Memoria

keep=False --> marca todas las filas como duplicados incluso la original (first/last)

head(6) --> muestra 6 registros

df_ventas.columns --> obtiene los rotulos de las columnas

list --> transforma esos nombres en una lista

sort_values --> se ordena filas duplicadas a través de esa lista

de esta manera es más rápido ubicar los duplicados


## La consigna pide revisar los valores para detectar posibles anomalías o rangos extraños
## Resumen estadístico de las columnas numéricas en ventas

In [ ]:

df_ventas.describe()

,id_venta,cantidad
count,3035.000000,3033.000000
mean,1499.851400,6.496538
std,866.465379,3.457250
min,1.000000,1.000000
25%,748.500000,3.000000
50%,1502.000000,7.000000
75%,2249.500000,9.000000
max,3000.000000,12.000000


## Ayuda Memoria



id_venta:

discrepancia count/max --> duplicados

min/mediana/max --> 1/~1500/3000 --> distribucion secuencial 1...3000

cantidad:

count/3035/3033 --> 2 registros nulos --> etapa limpieza (imputarse o eliminarse)

min/max --> 1/12 --> No hay ceros ni valores negativos --> comprar 1 o a lo sumo hsta 12 productos (razonable)

mean --> promedio de unidades por compra 6 (razonable)

25%/75% --> el 50 % de las ventas involucra entre 3 y  9 unidades (sin dispersiones extravagantes)

## Variables categóricas (texto) --> df_ventas
## Valores únicos/Valores más repetidos
## Resumen estadístico para variables de tipo texto (categóricas) en ventas

In [ ]:

df_ventas.describe(include=['object', 'str'])




,producto,precio,fecha_venta,categoria
count,3035,3033,3035,3035
unique,30,2590,364,3
top,Lámpara de mesa,$76.32,06/04/2024,Decoración
freq,181,5,24,1015


## Ayuda memoria

df_ventas.describe(include=['object']) --> sale un warning

El tipo de dato de texto tradicional (object) está migrando hacia el tipo específico de texto ('str').

Conviene ser explícito indicando ambos tipos: 'object' y 'str':

df_ventas.describe(include=['object', 'str'])

## Ayuda memoria
    
count--> cantidad de registros no nulos. Observar: 3035/3033 --> 2 registros nulos precio

unique--> valores distintos --> 30 productos/3 categorías * 364 fechas distintas (~1 año)

                            --> 2590 precios distintos
                            
top --> moda (valor frecuente) --> Lámpara de mesa (producto más vendido)
                           
                               --> Precio más repetido $ 76.32
                               
                               --> Día con mayor ventas 06/04/24
                               
                               --> Categoría Top Decoración
                               
freq --> frecuencia del valor top --> "Lámpara de mesa" aparece registrada 181 veces

                                  --> El precio "$76.32" se repitió 5 veces
                                  
                                  --> El "06/04/2024" hubo 24 ventas
                                  
                                  --> Decoración 1015 ventas (~ 1/3 del total)

## Ayuda memoria 
precio --> object/str : texto -->"$" --> fase de limpieza: float (calcular valores estadísticos)

## Limpieza y transformación de datos

## Ayuda memoria

Diagnóstico (df_ventas):

35 duplicados

valores nulos: 2(cantidad)/2(precio)

precio --> "$"

fecha_venta --> "texto"



## Eliminar las 35 filas duplicadas--> df_ventas

In [ ]:

df_ventas = df_ventas.drop_duplicates()

# Verificar que el total de filas ahora sea exactamente 3000
print("Filas y columnas actuales de df_ventas:", df_ventas.shape)
print("Duplicados restantes:", df_ventas.duplicated().sum())

Filas y columnas actuales de df_ventas: (3000, 6)
Duplicados restantes: 0


## Quitar el símbolo '$' y convertir a número flotante

In [ ]:

df_ventas['precio'] = df_ventas['precio'].str.replace('$', '', regex=False)
df_ventas['precio'] = pd.to_numeric(df_ventas['precio'], errors='coerce')

# Verificar el nuevo tipo de dato y las primeras filas
print("Nuevo tipo de dato de precio:", df_ventas['precio'].dtype)
df_ventas[['producto', 'precio']].head()

Nuevo tipo de dato de precio: float64


,producto,precio
0,Cuadro decorativo,69.94
1,Lámpara de mesa,105.10
2,Secadora,97.96
3,Heladera,114.35
4,Secadora,106.21


## Ayuda Memoria

## df_ventas['precio'] = df_ventas['precio'].str.replace('$', '', regex=False)

## df_ventas['precio']:

Selecciona la columna precio, que inicialmente se leyó como texto (object / string) porque contenía el símbolo de moneda (por ejemplo: "$1500.50").

## .str:

Es el accesor de cadenas de Pandas. Le indica a Pandas: "voy a aplicar una función de texto fila por fila a cada valor de esta columna". Sin .str, Pandas no sabría cómo manipular el texto dentro de una Serie.

## .replace('$', '', regex=False):

'$': El texto que querés buscar.

'': El texto por el cual lo reemplazás (en este caso, nada / vacío, es decir, lo estás borrando).

## regex=False: 

En expresiones regulares (regex), el símbolo $ significa "final de la línea". Al poner regex=False, le aclarás a Pandas que trate al $ como un carácter literal, un signo pesos común y corriente, evitando errores o comportamientos inesperados.

## df_ventas['precio'] = ...:

Sobrescribe la columna original con los nuevos valores limpios (por ejemplo, transforma "$1500.50" en "1500.50").

## la columna 

sigue siendo texto (solo que ahora no tiene el signo $). Para poder hacer cuentas matemáticas (como calcular total_venta), necesitás convertirla a número flotante con .astype(float) o pd.to_numeric()

# Ayuda Memoria


## df_ventas['precio'] = pd.to_numeric(df_ventas['precio'], errors='coerce')

## Esta línea convierte la columna precio a un tipo numérico (float o int), pero con una red de seguridad fundamental para evitar que el programa se rompa si encuentra valores extraños.


## pd.to_numeric(...)
Es una función nativa de Pandas diseñada específicamente para transformar datos a formato numérico. A diferencia de .astype(float), esta función está pensada para manejar columnas "sucias" o con datos inconsistentes.

## errors='coerce' (La clave de la línea)
El parámetro errors define qué debe hacer Pandas si se topa con un dato que no se puede convertir a número (por ejemplo: un espacio en blanco " ", un texto como "N/A", "gratis", o un símbolo que no se haya limpiado).

Pandas tiene tres comportamientos posibles:

## errors='raise' (por defecto): 
Si hay un solo valor inválido, corta el programa y lanza un error.

## errors='ignore': 
Si falla, deja la columna tal como estaba (en formato texto) sin avisar.

## errors='coerce' (forzar): 
Si un valor no se puede transformar a número, lo convierte automáticamente en NaN (nulo o valor faltante) en lugar de romper el código.

## ¿Por qué lo usamos?
Robustez: Si en alguna fila había un dato corrupto o mal escrito, no detiene la ejecución del notebook.

Facilita la auditoría: Todos los valores problemáticos quedan unificados como NaN.

Prepara el terreno para el paso siguiente: Luego podés usar df_ventas['precio'].isna().sum() para contar cuántos datos fallaron y tratarlos limpiamente (por ejemplo, eliminándolos con .dropna(), como se hizo con las 2 filas descartadas).

## Convertir fecha_venta a tipo datetime

In [ ]:

df_ventas['fecha_venta'] = pd.to_datetime(df_ventas['fecha_venta'], dayfirst=True)

# Verificar el tipo de dato
print("Tipo de dato de fecha_venta:", df_ventas['fecha_venta'].dtype)
df_ventas['fecha_venta'].head()

Tipo de dato de fecha_venta: datetime64[us]


0   2024-01-02
1   2024-01-02
2   2024-01-02
3   2024-01-02
4   2024-01-02
Name: fecha_venta, dtype: datetime64[us]

## Ayuda memoria

dayfirst=True--> el día viene antes que el mes

## Contar valores nulos por columna en df_ventas

In [ ]:

df_ventas.isnull().sum()

id_venta       0
producto       0
precio         2
cantidad       2
fecha_venta    0
categoria      0
dtype: int64

## Importante:

Antes de decidir qué hacer con ellos (imputarlos o eliminarlos), inspeccionemos cuáles son esas 4 filas afectadas.

Aislar las filas con valores nulos

## Filtrar y mostrar las filas que tienen nulos en precio o cantidad

In [ ]:

df_ventas[df_ventas['precio'].isnull() | df_ventas['cantidad'].isnull()]

,id_venta,producto,precio,cantidad,fecha_venta,categoria
136,627,Elementos de cerámica,NaN,NaN,2024-01-17,Decoración
139,2171,Parlantes Bluetooth,NaN,NaN,2024-01-17,Electrónica


## Importante:

Los 4 nulos están concentrados en 2 filas.

Sin precio ni cantidad no podemos especular nada.

Esas 2 filas representan el 0.07 % del total de transacciones.

Conclusión: eliminarlas

## Eliminar las filas con valores nulos en df_ventas

In [ ]:

df_ventas = df_ventas.dropna()

# Verificar que ya no queden nulos y que el total de filas sea 2998
print("Valores nulos restantes:\n", df_ventas.isnull().sum())
print("\nDimensiones finales de df_ventas:", df_ventas.shape)

Valores nulos restantes:
 id_venta       0
producto       0
precio         0
cantidad       0
fecha_venta    0
categoria      0
dtype: int64

Dimensiones finales de df_ventas: (2998, 6)


## Importante:

df_ventas:

--> sin duplicados

--> sin valores nulos

--> con 2998 transacciones válidas 

--> los tipos de datos correctos

## Creación de la columna calculada total_venta

--> ingresos totales

--> ventas por categoría o por mes

--> ventas por mes

--> precio por cantidad

## Calcular el monto total de cada transacción

In [ ]:

df_ventas['total_venta'] = df_ventas['precio'] * df_ventas['cantidad']

# Redondear a 2 decimales y previsualizar
df_ventas['total_venta'] = df_ventas['total_venta'].round(2)
df_ventas[['producto', 'precio', 'cantidad', 'total_venta']].head()

,producto,precio,cantidad,total_venta
0,Cuadro decorativo,69.94,5.0,349.70
1,Lámpara de mesa,105.10,5.0,525.50
2,Secadora,97.96,3.0,293.88
3,Heladera,114.35,8.0,914.80
4,Secadora,106.21,4.0,424.84


## Auditoría y limpieza de df_clientes
## Diagnóstico de df_clientes

In [ ]:

print("--- Diagnóstico de df_clientes ---")
print("Dimensiones:", df_clientes.shape)
print("Duplicados:", df_clientes.duplicated().sum())
print("\nValores nulos por columna:\n", df_clientes.isnull().sum())
print("\nTipos de datos:\n", df_clientes.dtypes)

--- Diagnóstico de df_clientes ---
Dimensiones: (567, 5)
Duplicados: 0

Valores nulos por columna:
 id_cliente    0
nombre        0
edad          0
ciudad        0
ingresos      0
dtype: int64

Tipos de datos:
 id_cliente      int64
nombre            str
edad            int64
ciudad            str
ingresos      float64
dtype: object


## No requiere ninguna acción.

## Auditoría y limpieza de df_marketing
## Diagnóstico de df_marketing

In [ ]:

print("--- Diagnóstico de df_marketing ---")
print("Dimensiones:", df_marketing.shape)
print("Duplicados:", df_marketing.duplicated().sum())
print("\nValores nulos por columna:\n", df_marketing.isnull().sum())
print("\nTipos de datos:\n", df_marketing.dtypes)

--- Diagnóstico de df_marketing ---
Dimensiones: (90, 6)
Duplicados: 0

Valores nulos por columna:
 id_campanha     0
producto        0
canal           0
costo           0
fecha_inicio    0
fecha_fin       0
dtype: int64

Tipos de datos:
 id_campanha       int64
producto            str
canal               str
costo           float64
fecha_inicio        str
fecha_fin           str
dtype: object


# En df_marketing no hay duplicados ni nulos, pero las columnas fecha_inicio y fecha_fin figuran como texto (str)

In [ ]:
# Convertir fechas de marketing a tipo datetime
df_marketing['fecha_inicio'] = pd.to_datetime(df_marketing['fecha_inicio'], dayfirst=True)
df_marketing['fecha_fin'] = pd.to_datetime(df_marketing['fecha_fin'], dayfirst=True)

# Verificar los tipos de datos actualizados
print("Nuevos tipos de datos en df_marketing:")
print(df_marketing[['fecha_inicio', 'fecha_fin']].dtypes)

Nuevos tipos de datos en df_marketing:
fecha_inicio    datetime64[us]
fecha_fin       datetime64[us]
dtype: object


# Verificación global final

In [ ]:
print(f"df_ventas: {df_ventas.shape} registros limpios | Columnas: {list(df_ventas.columns)}")
print(f"df_clientes: {df_clientes.shape} registros limpios | Columnas: {list(df_clientes.columns)}")
print(f"df_marketing: {df_marketing.shape} registros limpios | Columnas: {list(df_marketing.columns)}")

df_ventas: (2998, 7) registros limpios | Columnas: ['id_venta', 'producto', 'precio', 'cantidad', 'fecha_venta', 'categoria', 'total_venta']
df_clientes: (567, 5) registros limpios | Columnas: ['id_cliente', 'nombre', 'edad', 'ciudad', 'ingresos']
df_marketing: (90, 6) registros limpios | Columnas: ['id_campanha', 'producto', 'canal', 'costo', 'fecha_inicio', 'fecha_fin']
